rfcs/01391_async_sdk.md

```diff
@@ -0,0 +1,308 @@
+library, with abstractions for file system access, sockets etc. The Oak SDK does
+not provide such functionalities to nodes, and the ABI exposes only a single
+blocking call:
+[`wait_on_channels`](https://project-oak.github.io/oak/oak_abi/doc/oak_abi/index.html).
+
+While this makes existing executors unsuitable for use in nodes, the small API
+surface suggests implementing a custom executor for the Oak SDK is tractable.
+
+## Overview
+
+Support for asynchronous Rust consists of the following components:
+
+- executor for futures that runs in an Oak node.
+- async
+  [event loop](https://project-oak.github.io/oak/sdk/doc/oak/fn.run_event_loop.html).
+- asynchronous primitives for reading from channels. These are tightly coupled
+  to the executor.
+- spawn functionality for concurrent execution of futures.
+
+Put together, these components allow developers to implement asynchronous nodes.
+The
+[`hello_world`](https://github.com/project-oak/oak/blob/main/examples/hello_world/module/rust/src/lib.rs)
+example node would look similar to:
+
+```rust
+oak::entrypoint!(oak_main => |_| {
+    // Grpc server setup ...
+
+    oak::run_event_loop(grpc_channel, handler);
+});
+
+async fn handler(requests: GrpcInvocations<HelloWorld>) {
+    // Initial setup
+    let translator = { /* ... */ };
+
+    requests.try_for_each_concurrent(
+        /* concurrency limit */ 2,
+        |request| async move {
+            match request {
+                HelloWorld::SayHello(request, response) => {
+                    let translation = translator.translate_async(TranslateRequest {
+                        text: request.greeting,
+                        from: "en".to_string(),
+                        to: "fr".to_string(),
+                    }).await?;
+                    response.send(HelloResponse {
+                        reply: format!("BONJOUR {}!", translated_greeting),
+                    })?;
+                }
+                // ...
+            }
+    }).await.expect("Error handling request");
+}
+```
+
+Proper client-side streaming
+([an open issue](https://github.com/project-oak/oak/issues/97)) becomes
+straightforward now that the incoming requests can be represented as a
+[`Stream`](https://docs.rs/futures/0.3.5/futures/stream/trait.Stream.html):
+
+```rust
+match request {
+    // ...
+    HelloWorld::BidiHello(requests, responses) => {
+        requests.for_each(|request| {
+            responses.send(HelloResponse {
+                reply: format!("HELLO {}!", request.greeting),
+            })?;
+        }).await;
+    }
+}
+```
+
+New functions defined on the existing
+[`Receiver`](https://project-oak.github.io/oak/sdk/doc/oak/io/struct.Receiver.html)
+allow asynchronous reads from channels:
+
+```rust
+impl<T: Decodable> Receiver<T> {
+    pub fn receive_async(&self) -> ChannelRead<T> { /* ... */ }
+    pub fn receive_stream(&self) -> ChannelReadStream<T> { /* ... */ }
+}
```

Comment:
I feel this should just implement `Stream`, similar to https://docs.rs/tokio/0.2.22/tokio/sync/mpsc/struct.Receiver.html , then the caller can decide whether to read one or more elements off the stream. Or is there any benefit in having these two methods here?

Reply:
While the async stuff lives in `experimental` it is nice to implement it this way, since we can't touch the implementation of the `Receiver` itself.

This is a good point though, I can't think of any other reason not to implement `Stream` directly on the `Receiver`